Capgen in SCM: Fix to allow for scheme subcycling.

dustinswales opened the pull request:
Overview
This PR contains changes to fix scheme subcycling in Capgen and extends the var_compatability_test to exercise subcycling. 
UPDATE: Added bugfix for suite-part list ordering (see change to ccpp_suite.py).

Description
Create local group variable for subcycle indexing.
Fix bug (`self.loop`  -> `self._loop`) in the Subcycle write phase, and in ccpp_datafile.

User interface changes?: No

Fixes: #632 
Fixes: #634 

Testing:
Added to var_compatibility_test to exercise feature.

dustinswales committed:
Bug fix and exercise change in testing.

dustinswales committed:
Omission from previous commit

dustinswales committed:
Merge branch 'bugfix/scheme_ordering_in_caps' of https://github.com/dustinswales/ccpp-framework into bugfix/local_var_names_in_caps

dustinswales committed:
Fix subcycling and add test

climbfuji commented on scripts/metavar.py:
```diff
@@ -1677,12 +1677,37 @@ def add_variable(self, newvar, run_env, exists_ok=False, gen_unique=False,
             # end if
         # end if
+        # Check if local_name exists in Group. If applicable, Create new 
+        # variable with uniquie name. There are two instances when new names are
```
```suggestion
        # variable with unique name. There are two instances when new names are
```

climbfuji commented on scripts/metavar.py:
```diff
@@ -1677,12 +1677,37 @@ def add_variable(self, newvar, run_env, exists_ok=False, gen_unique=False,
                                        context=newvar.context)
             # end if
         # end if
+        # Check if local_name exists in Group. If applicable, Create new 
```
```suggestion
        # Check if local_name exists in Group. If applicable, create new 
```

climbfuji commented on scripts/metavar.py:
```diff
@@ -1677,12 +1677,37 @@ def add_variable(self, newvar, run_env, exists_ok=False, gen_unique=False,
             if gen_unique:
                 new_lname = self.new_internal_variable_name(prefix=lname)
                 newvar = newvar.clone(new_lname)
+                # Local_name needs to be the local_name for the new
```
I am very confused by this comment. Should the first `Local_name` be `lname`?

dustinswales replied:
All instances of `lname` are replaced, there is no logic to keep the first instance. So for three instances of the same lname, we have `lname1`, `lname2`, and `lname3`, not `lname`, `lname1`, `lname2`.

dustinswales replied:
Also, this is part of #631 and does not impact the sub-cycling isssue

climbfuji commented on test/var_compatibility_test/var_compatibility_suite.xml:
```diff
@@ -5,6 +5,8 @@
     <scheme>effr_calc</scheme>
     <scheme>effr_post</scheme> 
-    <scheme>effr_diag</scheme>
+    <subcycle loop="1">
```
Would it make sense to test with more than one loop iteration?

dustinswales replied:
Done, although bare bone. See 1fe9f2e

dustinswales committed:
Remove sorting of suite part list

dustinswales committed:
Address reviewers comments

dustinswales committed:
Exercise subcycling in var_compatability test

climbfuji committed:
Merge branch 'develop' into bugfix/local_var_names_in_caps

dustinswales committed:
Merge branch 'develop' of https://github.com/NCAR/ccpp-framework into bugfix/scheme_subcycling

dustinswales committed:
Change one test scalar to type int.

dustinswales committed:
Add modification for CAM-SIMA

dustinswales committed:
Merge branch 'bugfix/local_var_names_in_caps' into bugfix/scheme_subcycling

dustinswales committed:
Merge branch 'develop' of https://github.com/NCAR/ccpp-framework into bugfix/scheme_subcycling

dustinswales committed:
Merge branch 'develop' of https://github.com/NCAR/ccpp-framework into bugfix/scheme_subcycling

dustinswales committed:
Omission from previous commmit

gold2718 commented:
This PR only seems to allow literal integers as a loop index, is that right? I tried adding a fix to allow entering a standard name instead (an NCAR requirement since these loop indices are run-time variables) but ran into:
```
Invalid local_name: scalar_var already registered, in /home/goldy/Projects/CAMDEN/ccpp_framework/test/var_compatibility_test/var_compatibility_suite.xml
```
which I thought was fixed in #631. This only happens if I use a standard name in the SDF. This is weird because the standard name has nothing to do with `scalar_var`. Any ideas?

Commit:
Permit variables as loop variables
A standard name can be used instead of a literal integer
Modified the var_compatibility test to test both a
literal integer subcycle and a loop variable subcycle.

gold2718 commented:
@dustinswales, I think I found the issues. I have submitted a PR to this PR branch with the fixes. The var_compatibility test now includes both a standard_name (variable) subcycle and a literal integer subcycle. As a bonus, they are nested!

dustinswales committed:
Merge pull request #5 from gold2718/scheme_subcycling

Permit variables as loop variables

dustinswales commented:
@gold2718 Thanks for adding this functionality! We have this hardcoded at the moment on the UFS side, but I could see us adopting this for suite extensibility.

gold2718 reviewed (approved):
Looks good to me!

dustinswales commented:
@peverwhee @climbfuji  Can you review this one when you get a chance?
